stateproof/builder.go

```diff
@@ -0,0 +1,470 @@
+	err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		if brnd%basics.Round(proto.StateProofInterval) < basics.Round(proto.StateProofInterval/2) {
+			roundSigs, err = getPendingSigsFromThisNode(tx)
+		} else {
+			roundSigs, err = getPendingSigs(tx)
+		}
+		return
+	})
+	if err != nil {
+		spw.log.Warnf("broadcastSigs: getPendingSigs: %v", err)
+		return
+	}
+
+	for rnd, sigs := range roundSigs {
+		if rnd > brnd {
+			// Signature is for later block than brnd.  This could happen
+			// during catchup or testing.  The caller's loop will eventually
+			// invoke this function with a suitably high brnd.
+			continue
+		}
+
+		for _, sig := range sigs {
+			// Randomize which sigs get broadcast over time.
+			addr64 := binary.LittleEndian.Uint64(sig.signer[:])
+			if addr64%(proto.StateProofInterval/2) != uint64(brnd)%(proto.StateProofInterval/2) {
+				continue
+			}
+
+			sfa := sigFromAddr{
+				SignerAddress: sig.signer,
+				Round:         rnd,
+				Sig:           sig.sig,
+			}
+			err = spw.net.Broadcast(context.Background(), protocol.StateProofSigTag,
+				protocol.Encode(&sfa), false, nil)
+			if err != nil {
+				spw.log.Warnf("broadcastSigs: Broadcast for %d: %v", rnd, err)
+			}
+		}
+	}
+}
+
+func lowestRoundToRemove(currentHdr bookkeeping.BlockHeader) basics.Round {
+	proto := config.Consensus[currentHdr.CurrentProtocol]
+	nextStateProofRnd := currentHdr.StateProofTracking[protocol.StateProofBasic].StateProofNextRound
+	if proto.StateProofInterval == 0 {
+		return nextStateProofRnd
+	}
+
+	recentRoundOnRecoveryPeriod := basics.Round(uint64(currentHdr.Round) - uint64(currentHdr.Round)%proto.StateProofInterval)
+	oldestRoundOnRecoveryPeriod := recentRoundOnRecoveryPeriod.SubSaturate(basics.Round(proto.StateProofInterval * proto.StateProofRecoveryInterval))
+	// we add +1 to this number since we want exactly StateProofRecoveryInterval elements in the history
+	oldestRoundOnRecoveryPeriod++
+
+	var oldestRoundToRemove basics.Round
+	if oldestRoundOnRecoveryPeriod > nextStateProofRnd {
+		oldestRoundToRemove = oldestRoundOnRecoveryPeriod
+	} else {
+		oldestRoundToRemove = nextStateProofRnd
+	}
+	return oldestRoundToRemove
+}
+
+func (spw *Worker) deleteOldSigs(currentHdr bookkeeping.BlockHeader) {
+	oldestRoundToRemove := lowestRoundToRemove(currentHdr)
+
+	err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		return deletePendingSigsBeforeRound(tx, oldestRoundToRemove)
+	})
+	if err != nil {
+		spw.log.Warnf("deletePendingSigsBeforeRound(%d): %v", oldestRoundToRemove, err)
+	}
+}
+
+func (spw *Worker) deleteOldBuilders(currentHdr bookkeeping.BlockHeader) {
+	oldestRoundToRemove := lowestRoundToRemove(currentHdr)
+
+	spw.mu.Lock()
+	defer spw.mu.Unlock()
+
+	for rnd := range spw.builders {
+		if rnd < oldestRoundToRemove {
+			delete(spw.builders, rnd)
+		}
+	}
+}
+
+func (spw *Worker) tryBuilding() {
+	spw.mu.Lock()
+	defer spw.mu.Unlock()
+
+	for rnd, b := range spw.builders {
```

Comment:
Did you ever sort the builders in ascending order? Otherwise the transaction pool will reject out-of-order transactions, which is wasted effort if you ever send more than one.

Reply:
Yes. we have a PR for that, but it didn't yield any benefit so we pushed it back.
Since we have more time we would add it.